[PT FE] Add aten::atan2

rghvsh opened the pull request:
### Details:
 - *Adds aten::atan2 operator*

### Tickets:
 - Closes #20575

mvafin commented on tests/layer_tests/pytorch_tests/test_atan2.py:
```diff
@@ -0,0 +1,46 @@
+
+    def create_atan2_net(self, input_shape, input_type):
+        self.input_type = input_type
+        tf.compat.v1.reset_default_graph()
+        # Create the graph and model
+        with tf.compat.v1.Session() as sess:
+            y = tf.compat.v1.placeholder(input_type, input_shape, 'y')
+            x = tf.compat.v1.placeholder(input_type, input_shape, 'x')
+            tf.raw_ops.Atan2(y=y, x=x)
+            tf.compat.v1.global_variables_initializer()
+            tf_net = sess.graph_def
```
Here you create tensorflow model. Please create pt model

mvafin commented on src/frontends/pytorch/src/op/atan2.cpp:
```diff
@@ -0,0 +1,72 @@
+
+    // handle the fifth condition : x=0 && y<0
+    auto cond4 = make_shared<v1::LogicalAnd>(is_x_zero, is_y_negative);
+    auto const_minus_two = create_same_type_const_scalar<int32_t>(x, -2);
```
You copied this from tf fe, but this helper doesn't exist in pt fe, so this code will not build.

mvafin reviewed (changes requested):
Please check that code builds before submitting

rghvsh commented:
Hey @mvafin! I wrote this code for tffe this was a first draft will work on this

rghvsh commented:
Hi! @mvafin made some changes is this better?

Thanks
Raghav

mvafin commented on src/frontends/pytorch/src/op/atan2.cpp:
```diff
@@ -0,0 +1,72 @@
+    auto pi_div_minus_two = make_shared<v1::Divide>(const_pi, const_minus_two);
+    result = make_shared<v1::Select>(cond4, pi_div_two, result);
+
+    set_node_name(context.get_name(), result.get_node_shared_ptr());
```
There is no such helper in PyTorch Frontend. Does your code build?

mvafin commented on src/frontends/pytorch/src/op/atan2.cpp:
```diff
@@ -0,0 +1,72 @@
+namespace op {
+
+OutputVector translate_atan2_op(const NodeContext& context) {
+    default_op_checks(context, 2, {"Atan2"});
```
This helper is used in TF FE, in PT fE we have a different helper. Please change

mvafin commented on src/frontends/pytorch/src/op/atan2.cpp:
```diff
@@ -0,0 +1,72 @@
+    auto x = context.get_input(1);
+
+    // handle the first condition : x>0
+    auto div_y_x = make_shared<v1::Divide>(y, x);
```
Please call context.mark_node for each node you create

rghvsh replied:
sure

mvafin reviewed (changes requested):
Also

rghvsh commented:
Hey @mvafin made the recommended changes please review once.

mvafin commented on tests/layer_tests/pytorch_tests/test_atan2.py:
```diff
@@ -0,0 +1,45 @@
+
+class TestAtan2(PytorchLayerTest):
+    class TestAtan2(PytorchLayerTest):
+    def _prepare_input(self, inputs_info):
```
there is no `inputs_info` for pytorch models, you can just just create tensors here

mmikolajcz commented on src/frontends/pytorch/src/op_table.cpp:
```diff
@@ -341,6 +342,8 @@ const std::map<std::string, CreatorFunction> get_supported_ops_ts() {
         {"aten::asinh_", op::inplace_op<op::translate_1to1_match_1_inputs<opset10::Asinh>>},
         {"aten::atan", op::optional_out<op::translate_1to1_match_1_inputs_with_fp32_type_alignment<opset10::Atan>, 1>},
         {"aten::atan_", op::inplace_op<op::translate_1to1_match_1_inputs<opset10::Atan>>},
+        {"aten::atan2", op::translate_atan2_op},
```
I experienced some build issues there, either you need to remove `_op` part from function name (also in translation) or add this `_op` in `OP_CONVERTER(translate_atan2);`. Generally, in PyTorch Frontend we don't use _op at the end so it may be better to stick to existing name convention.

mmikolajcz commented on tests/layer_tests/pytorch_tests/test_atan2.py:
```diff
@@ -0,0 +1,45 @@
+from pytorch_layer_test_class import PytorchLayerTest
+
+class TestAtan2(PytorchLayerTest):
+    class TestAtan2(PytorchLayerTest):
```
```suggestion
```
Duplicated class declaration, please remove

mmikolajcz commented on src/frontends/pytorch/src/op/atan2.cpp:
```diff
@@ -0,0 +1,73 @@
+#include "pt_framework_node.hpp"
+#include "utils.hpp"
+
+using namespace std;
```
It seem that `using namespace std;` is used just for `make_shared` - Please remove it and use regular `std::make_shared` to follow convention from other translations or just replace it by `using std::make_shared`. By using whole std namespace, we risk name/class collision in future

mmikolajcz commented on src/frontends/pytorch/src/op/atan2.cpp:
```diff
@@ -0,0 +1,73 @@
+    auto x = context.get_input(1);
+
+    // handle the first condition : x>0
+    auto div_y_x = context.mark_node(make_shared<v1::Divide>(y, x));
```
Inputs in torch often have mixed element types and require promotion to common type in order to allow OV ops to work correctly. Please check how this operator behaves in torch with different input types (like mixed i32, f32, f16) and what output type is. In OV we have helper functions, like `align_promote_types` and others from utils

rghvsh replied:
Can a solution to this be we convert the input in whatever form it is to f32 and all the constants in the program are also of the same type? Something like:
auto x = context.mark_node(std::make_shared<v0::Convert>(context.get_input(0), element::f32));

mmikolajcz replied:
Not really, it will likely be bit tricky op to properly align input types. If I understand this op correctly:

1. First of, inputs need to be promoted to common type, like you would end up if you would simply torch.add those inputs.
2. Secondly, any integer types need to be converted to f32 (but not floating - they remain as is).

I think easiest way to do it would be after https://github.com/openvinotoolkit/openvino/pull/22770 would be merged.

To get inputs in common promoted types (point 1.) you could just call ` std::tie(y, x) = get_inputs_with_promoted_types(context, 0, 1);` (look into changes in add.cpp for more details)

If you end up with int datatype as result  (point 2.) you could do it by promoting type to dummy const with rank 0 and mark it as PyScalar (last argument set to `true`):
```
auto dummy_const = context.mark_node(ov::op::v0::Constant::create(element::f32, Shape({}), {0.5}))->output(0);
align_eltwise_input_types(context, x, dummy_const, false, true);
```

mmikolajcz commented on src/frontends/pytorch/src/op/atan2.cpp:
```diff
@@ -0,0 +1,73 @@
+
+    // handle the second condition : x<0 && y>=0
+    auto const_pi = v0::Constant::create(element::i32, Shape{}, {std::atan(1.0)*4});
+    auto is_x_negative = context.mark_node(make_shared<v1::Less>(x, const_zero));
```
Less operator requires to have both inputs to have same type. This will fail if x would have type other than i32. Probably there would be some more issues in other ops

mmikolajcz commented on tests/layer_tests/pytorch_tests/test_atan2.py:
```diff
@@ -0,0 +1,45 @@
+        y_shape = inputs_info['y']
+        x_shape = inputs_info['x']
+        inputs_data = {}
+        inputs_data['y'] = np.random.rand(*y_shape).astype(self.input_type) - np.random.rand(*y_shape).astype(self.input_type)
```
Please consider adding some manually prepared tensor values (instead rand) to be sure that each run contains values for each possible plane of atan2, plus some edge cases, like case where x and y are 0, have higher values or are at the edge of each plane.

rghvsh replied:
Here I was trying that we can have two cases:
1. In which term 1 is bigger than term 2 so result positive
2. In which term 2 is bigger than term2 so result negative
This will make random positive/negative value for x and y so we get coverage of all cases.
Will include edge cases like
 x = y = 0

rghvsh commented:
Hi! @mvafin and @mmikolajcz Thank you for the review will make the changes

rghvsh commented:
Hi! @mvafin and @mmikolajcz Thank you for the review will make the changes

rghvsh commented:
Hi! @mvafin and @mmikolajcz Thank you for the review will make the changes

mvafin commented on tests/layer_tests/pytorch_tests/test_atan2.py:
```diff
@@ -0,0 +1,40 @@
+from pytorch_layer_test_class import PytorchLayerTest
+
+class TestAtan2(PytorchLayerTest):
+    def _prepare_input(self, x, y, dtype=None):
```
`_prepare_input` doesn't have this signature, if you want to use `kwargs` you need to provide them to `_test` call.

rghvsh commented:
Hi! @mvafin and mmikolajcz made changes to the test and translation please check.

Thanks
Raghav

mlukasze commented:
hey @rghvsh while you are waiting for next round of the review, could you please fix at least clang check?
and take a look for rest of red ones (they may be, but may be not something you should correct as well)

mlukasze commented:
@mvafin & @mmikolajcz I would like to ask to finalize this PR with our contributor. It took some time and effort, it's worth to have it for upcoming release.

mmikolajcz commented on src/frontends/pytorch/src/op/atan2.cpp:
```diff
@@ -0,0 +1,100 @@
+namespace op {
+
+OutputVector translate_atan2(const NodeContext& context) {
+    num_inputs_check(context, 3, 3);
```
```suggestion
    num_inputs_check(context, 2, 3);
```
Looking on your code, op should contain either 2 or 3 inputs instead 3

mmikolajcz commented on src/frontends/pytorch/src/op/atan2.cpp:
```diff
@@ -0,0 +1,100 @@
+OutputVector translate_atan2(const NodeContext& context) {
+    num_inputs_check(context, 3, 3);
+    // "aten::atan2.out(Tensor input,Tensor other, *,Tensor(a!) out) → Tensor(a!)"
+    auto y = context.get_input(0);
+    auto x = context.get_input(1);
+    std::tie(y, x) = get_inputs_with_promoted_types(context, 0, 1);
```
```suggestion
    Output<Node> y;
    Output<Node> x;
    std::tie(y, x) = get_inputs_with_promoted_types(context, 0, 1);
```
If you use `get_inputs_with_promoted_types` you don't need to retrieve them using `context.get_input(...)`.
Plus please pull changes from master branch, those helpers were added recently, I don't see those changes on your fork.

mmikolajcz commented on src/frontends/pytorch/src/op/atan2.cpp:
```diff
@@ -0,0 +1,100 @@
+    align_eltwise_input_types(context, x, dummy_const, false, true);
+
+    // handle the first condition : x>0
+    align_eltwise_input_types(context, y, x, false, true);
```
This `align_eltwise_input_types` is unnecessary - inputs at his point already should be aligned

mmikolajcz commented on src/frontends/pytorch/src/op/atan2.cpp:
```diff
@@ -0,0 +1,100 @@
+
+    // handle the second condition : x<0 && y>=0
+    auto const_pi = v0::Constant::create(element::f32, Shape{}, {std::atan(1.0)*4});
+    align_eltwise_input_types(context, x, const_pi, false, true);
```
In such simple conversions of const to type of nodes, please use op ConvertLike instead align_eltwise_input_types. Some conversions are redundant and could be called just once, please remove them

mmikolajcz commented on src/frontends/pytorch/src/op/atan2.cpp:
```diff
@@ -0,0 +1,100 @@
+    result = context.mark_node(std::make_shared<v1::Select>(cond4, pi_div_two, result));
+
+    // check whether out tensor is given
+    if(!context.input_is_none(2) && context.get_input_size == 3){
+    out_tensor = context.get_input(2);
+    // dtype is inherited from out tensor in input
+    auto result_out = context.mark_node(std::make_shared<v1::ConvertLike>(result,out_tensor);
+
+    return {result_out};
```
In this code, out_tensor was not declared (missing auto) and in ConvertLike line parentheses are not closed causing build to fail. To handle out tensor, it would also need to use context.mutate_input(...).
If out is always on input with index 2, maybe it would be better to use `op::optional_out` helper in op_table.cpp?
You would need to modify `{"aten::atan2", op::translate_atan2},` to `{"aten::atan2", op::optional_out<op::translate_atan2, 2>},`. You can look at `aten::rsqrt` as an example.

rghvsh replied:
Hi! @mmikolajcz thanks, sure will take a look

mmikolajcz commented on tests/layer_tests/pytorch_tests/test_atan2.py:
```diff
@@ -0,0 +1,69 @@
+        
+    @pytest.mark.nightly
+    @pytest.mark.precommit
+    @pytest.mark.parametrize("dtype", [None, "float32", "float64", "int32", "int64", "int8", "uin8"])
```
```suggestion
    @pytest.mark.parametrize("dtype", [None, "float32", "float64", "int32", "int64", "int8", "uint8"])
```
Typo.
Looking at this test, it seem that `dtype` will affect only `out` tensor. Please add conversion to types for x and y input, either directly in model or at _prepare_inputs stage.
Also please consider adding few test for cases where x and y types are mixed - this op has quite complex type inference, it would be good to cover it by few simple test cases

rghvsh replied:
Hi @mmikolajcz 
` @pytest.mark.parametrize("y, x", [(0, 1.5), (0, 0), (1.25, -5), (1, 10), (-1, -5.5), (-1, -5), (1.25, -5.5), (1.9, 2.9), [10, 9.9]])`
is this fine or in _prepare_inputs should I add two parameters _prepare_inputs(self, y, x, dtype1, dtype2) where they are two different types.

mmikolajcz commented on tests/layer_tests/pytorch_tests/test_atan2.py:
```diff
@@ -0,0 +1,69 @@
+    @pytest.mark.precommit
+    @pytest.mark.parametrize("dtype", [None, "float32", "float64", "int32", "int64", "int8", "uin8"])
+    @pytest.mark.parametrize(
+        "start,end,steps", [(0, 1), (0, 0), (1, -5), (1, 10), (-1, -5), (-1, -5), (1.25, -5.5)]
```
Parameters start, end and steps aren't used by function, maybe you meant `"y, x"`?

mmikolajcz commented on tests/layer_tests/pytorch_tests/test_atan2.py:
```diff
@@ -0,0 +1,69 @@
+class TestAtan2(PytorchLayerTest):
+    def _prepare_input(self, y, x, ref_dtype=None):
+        inputs = [np.array(y).astype(ref_dtype) - np.array(y).astype(ref_dtype), np.array(x).astype(ref_dtype) - np.array(x).astype(ref_dtype)]
+        if ref_dtype:
+            inputs.append(np.zeros(1).astype(ref_dtype))
```
This ref_dtype input seem to not be used

mvafin commented:
@rghvsh Could you at least verify that your code builds and passes your tests before asking for next review iteration?

mlukasze commented:
@mmikolajcz & @mvafin please retake a review

mvafin commented on src/frontends/pytorch/src/op/atan2.cpp:
```diff
@@ -0,0 +1,100 @@
+#include "openvino/op/range.hpp"
+#include "openvino/op/select.hpp"
+#include "openvino/op/subtract.hpp"
+#include "pt_framework_node.hpp"
```
I don't think you need this include

mvafin commented on src/frontends/pytorch/src/op/atan2.cpp:
```diff
@@ -0,0 +1,100 @@
+    std::tie(y, x) = get_inputs_with_promoted_types(context, 0, 1);
+    auto dummy_const = context.mark_node(ov::op::v0::Constant::create(element::f32, Shape({}), {0.5}))->output(0);
+    // align input types
+    align_eltwise_input_types(context, x, dummy_const, false, true);
```
What you want to achieve with this? You called `get_inputs_with_promoted_types` which will promote types to closest common, which can be integer if both were integer. If you want to promote all to closest float type, then you would need to switch the order of calls: first call `align_eltwise_input_types` to x and dummy fp32 const and then call `get_inputs_with_promoted_types` to promote x and y to closest common type.

mmikolajcz replied:
I don't think this solution would work - `get_inputs_with_promoted_types` collects input using `get_input(...)` but doesn't mutate/overwrite them as far as I'm aware, so `align_eltwise_input_types` outcome would probably be ignored. 
Instead, adding `ConvertLike(y, x)` to this code should fix it. Another approach would be one you mentioned, but with second `align_eltwise_input_types(context, y, x, is_python_scalar_input(context, 0), is_python_scalar_input(context, 1)` instead `get_inputs_with_promoted_types`

mvafin commented on src/frontends/pytorch/src/op/atan2.cpp:
```diff
@@ -0,0 +1,100 @@
+    result = context.mark_node(std::make_shared<v1::Select>(cond4, pi_div_two, result));
+
+    // check whether out tensor is given
+    if(!context.input_is_none(2) && context.get_input_size() == 3) {
+        context.mutate_input(2, result);            
+    }
```
You used `optional_out` decorator in op_table, so you do not need to call `mutate_input` here.

mvafin commented on tests/layer_tests/pytorch_tests/test_atan2.py:
```diff
@@ -0,0 +1,67 @@
+        
+    @pytest.mark.nightly
+    @pytest.mark.precommit
+    @pytest.mark.parametrize("dtype", [None, "float32", "float64", "int32", "int64", "int8", "uint8"])
```
Also test mixed types for x and y

mvafin reviewed (changes requested):
Please also fix code style

rghvsh commented:
Hi! @mvafin @mmikolajcz I think i've fixed build issues and added test for different types in x and y. I had some questions:
 As there is `optional_out` decorator I don't need to use mutate_input or apply the if block and return result straight away?
 How to fix the `get inputs with promoted types`?

Thanks,
Raghav

rghvsh commented:
Hi @mvafin @mmikolajcz! Made the changes. Last time for the CI I guess the cmake did not configure properly.

mvafin commented on src/frontends/pytorch/src/op/atan2.cpp:
```diff
@@ -0,0 +1,100 @@
+    // Check whether inputs present
+    num_inputs_check(context, 2, 3);
+    // "aten::atan2.out(Tensor input,Tensor other, *,Tensor(a!) out) → Tensor(a!)"
+    Output<Node> y;
+    Output<Node> x;
+
+    // tie inputs together
+    std::tie(y, x) = get_inputs_with_promoted_types(context, 0, 1);
```
```suggestion
    Output<Node> x = context.get_input(0);
    Output<Node> y = context.get_input(1);
```

mvafin commented on src/frontends/pytorch/src/op/atan2.cpp:
```diff
@@ -0,0 +1,100 @@
+    align_eltwise_input_types(context, x, dummy_const, false, true);
+
+    // align input types of y, x     
+    align_eltwise_input_types(context, y, x, false, true);
```
```suggestion
    align_eltwise_input_types(context, x, y, is_python_scalar_input(context, 0), is_python_scalar_input(context, 1));
```

rghvsh commented:
Hi @mvafin @mmikolajcz! Made the changes

mlukasze commented:
build_jenkins

mvafin commented:
Still same error:
```
    def test_atan2_with_out(self, dtype1, dtype2, use_out, y, x, ie_device, precision, ir_version):
        self._test(
>           *self.create_model(dtype=dtype1, use_out=use_out),
            ie_device,
            precision,
            ir_version,
            kwargs_to_prepare_input={"y": y, "x": x}
        )
E       TypeError: TestAtan2.create_model() got an unexpected keyword argument 'dtype'
```

mlukasze commented:
hey @rghvsh will you find a time to address Maxim's requests?
